### DISCUSSION TITLE
Is there a way to mount a USB MSC on a specific FLASH segment?

### DISCUSSION THREAD

**Discussion by HensonWoo** [micropython/micropython#13489] [STM32 / Pyboard]
Hello, I hope you are doing well

*My developement board is STM32 NUCLEO_L476RG.

I have a question of flash parition, my intention is to divide the internal flash memory into two partitions, /flash and /data.
As described in the link below:
https://docs.micropython.org/en/latest/reference/filesystem.html#hybrid-stm32

But my use case is a little different.
The /flash is for python files (eg. boot.py, main.py, etc...) and its filesystem is littlefs
The /data is for log csv file and USB MSC and its filesystem is FAT
So, my questions are
1. How to seperate the flash parition in the stm32l476xg.ld and mpconfigboard.h
2. How to specify that the file system format on /flash is littlefs and /data is FAT
3. How to set the USB MSC to mount on the /data

Thanks in advance for your help.

**Comment by nspsck** [comment]
Hi, I hope the following information are useful to you.

> 3. How to set the USB MSC to mount on the /data

The following operations **are going to destroy your data**. The approach is simple, since the `stm32 `port mounts the `pyb.Flash()` starting from 0, and you only need 2 partitions, you can just format the begining as your `/data` and the end as your `/flash`. 

```py
import os, pyb
os.umount('/flash')
SIZE_IN_KIB = 256
pdata = pyb.Flash(start=0, len=SIZE_IN_KIB *1024)
pcode = pyb.Flash(start=SIZE_IN_KIB *1024)
# Format pdata section to FAT.
os.VfsFat.mkfs(pdata)
# Format pcode section to LittleFS2.
os.VfsLfs2.mkfs(pcode)
# Mount pcode section to /flash.
os.mount(pcode, '/flash')
# Mount pdata section to /data.
os.mount(pdata, '/data')
# This changes the working directory to /flash.
os.chdir('/flash')
```

> 2. How to specify that the file system format on /flash is littlefs and /data is FAT

Hopefully the comments in the provided code made things a little more clear.

> 1. How to seperate the flash parition in the stm32l476xg.ld and mpconfigboard.h

As you can see above, the described behavior can be achieved without editing the `.ld` file. However, if you want to learn more about how the '.ld' files, please view [the link](https://github.com/nspsck/STM32F411CEU6_BlackPill_Micropython?tab=readme-ov-file#storage-modification). Though it targets `stm32f411ce`, but you can still get a general idea on what those `.ld` files does. In your case, the NUCLEO_L476RG, the `SRAM2` is equivalent to the `FS_CACHE`. And you really do not need to worry about it at all. All you can adjust is the firmware size and sorage size respecting the fact that each sector is of the size of 2K.

If all these were not enough to suit your needs, the [LEGO_HUB_NO6](https://github.com/micropython/micropython/tree/master/ports/stm32/boards/LEGO_HUB_NO6) port could be a good case to study on.
